java-checks/src/main/java/org/sonar/java/checks/HardcodedURICheck.java

```diff
@@ -67,20 +77,83 @@ public class HardcodedURICheck extends IssuableSubscriptionVisitor {
   private static final Pattern URI_PATTERN = Pattern.compile(URI_REGEX + "|" + LOCAL_URI + "|" + DISK_URI + "|" + BACKSLASH_LOCAL_URI);
   private static final Pattern VARIABLE_NAME_PATTERN = Pattern.compile("filename|path", Pattern.CASE_INSENSITIVE);
   private static final Pattern PATH_DELIMETERS_PATTERN = Pattern.compile("\"/\"|\"//\"|\"\\\\\\\\\"|\"\\\\\\\\\\\\\\\\\"");
+  private static final Pattern RELATIVE_URI_PATTERN = Pattern.compile("^(/[\\w-+!*.]+){1,2}");
+
+
+  // we use these variables to track when we are visiting an annotation
+  private final Deque<AnnotationTree> annotationsStack = new ArrayDeque<>();
+
+  private record IdentifierData(Symbol symbol, String identifier) {
+  }
+
+  private final List<IdentifierData> identifiersUsedInAnnotations = new ArrayList<>();
+
+  private record VariableData(Symbol symbol, String identifier, ExpressionTree initializer) {
+  }
+
+  private final List<VariableData> hardCodedUri = new ArrayList<>();
+
+  @Override
+  public void setContext(JavaFileScannerContext context) {
+    super.setContext(context);
+    annotationsStack.clear();
+    identifiersUsedInAnnotations.clear();
+    hardCodedUri.clear();
+  }
+
+  @Override
+  public void leaveFile(JavaFileScannerContext context) {
+    // now, we know all variable that are used in annotation so we can report issues
```

Comment:
Did you consider doing it in one loop? It might be easier to read:
```
    // now, we know all variable that are used in annotation so we can report issues
    Set<Symbol> idSymbols = new HashSet<>();
    Set<String> idNamesWithSemantic = new HashSet<>();
    Set<String> idNamesWithoutSemantic = new HashSet<>();

    for(IdentifierData i : identifiersUsedInAnnotations) {
      if(i.symbol().isUnknown()) {
        idNamesWithoutSemantic.add(i.identifier());
      } else {
        idSymbols.add(i.symbol());
        idNamesWithSemantic.add(i.identifier());
      }
    }
```

Reply:
I did not think about not using the stream, yes in this case using a loop is better